### ISSUE TITLE
Inferred nullability of type parameters doesn't take nullability learnt from contract annotations into account

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#33385]
**Version Used**: 

Microsoft Visual Studio Enterprise 2019 Preview
Version 16.0.0 Preview 2.1
VisualStudio.16.Preview/16.0.0-pre.2.1+28529.54
Microsoft .NET Framework
Version 4.7.03056

**Steps to Reproduce**:

Compile the following code:
```cs
#nullable enable
using System.Runtime.CompilerServices;

class C
{
  void Test1(string? s1)
  {
    var s2 = M(s1, s1);
    s1.ToString(); // no warnings, Roslyn knows that s1 cannot be null after M(s1, s1) call
    s2.ToString(); // false warning, the second argument is knonwn to be not null, therefore T is non-nullable string
    s2 = null;
  }

  static extern T M<T>([EnsuresNotNull] string? s, T t);
}

namespace System.Runtime.CompilerServices
{
  class EnsuresNotNullAttribute : System.Attribute { }
}
```

**Expected Behavior**:
No warnings for the dereference of `s2`. Warning for assigning `null` to `s2`.
`s1` is known to be not null after the call. Therefore `C.M'T` should be inferred to be a non-nullable string and `s2` must be inferred to be a non-nullable string.

**Actual Behavior**:
`s2` is inferred to be a nullable string which is incorrect. No warnings for assigning it with `null` and false warnings on its derferences.

**Notes**
Similar example with explicit non-nullable value produces false warning for nullability mismatch in the variable's initializer
`string s2 = M(s1, s1);`

If I split assertion and type arguments inference into separate calls Roslyn is able to produce correct warnings e.g.
```cs
#nullable enable
using System.Runtime.CompilerServices;

class C
{
  void Test1(string? s1)
  {
    AssertionOnly(s1);
    var s2 = InferenceOnly(s1);
    s2.ToString(); // no warnings
    s2 = null; // cannot assign null to a non-nullable target
  }

  static extern void AssertionOnly([EnsuresNotNull] string? s);
  static extern T InferenceOnly<T>(T t);
}

namespace System.Runtime.CompilerServices
{
  class EnsuresNotNullAttribute : System.Attribute { }
}
```

**Comment by gafter** [comment]
I believe the OP is incorrect.  `T` should be inferred to `string?`.  The second argument to the method call is of type `string?`.  It is not guaranteed to be non-null until after the call.  If the second argument to `M` were some complex expression involving the use of the variable `s1`, it would not be safe to assume that `s1` is non-null while evaluating that argument.

**Comment by TessenR** [comment]
@gafter Could you please explain the difference regarding the second argument evaluation if the method is just a regular non-generic assertion method?

> The second argument to the method call is of type `string?`. It is not guaranteed to be non-null until after the call. If the second argument to `M` were some complex expression involving the use of the variable `s1`, it would not be safe to assume that `s1` is non-null while evaluating that argument.

The same is true for any assertion method with multiple arguments e.g.
```cs
#nullable enable
using System.Runtime.CompilerServices;

class C
{
  void Test1(string? s1)
  {
    string? s2;
    M(s1, s2 = s1, s1.ToString()); // it's not safe to assume s1 is not null yet, hence warning for s1.ToString()
    s2.ToString(); // no warnings
  }

  static extern string M([EnsuresNotNull] string? s1, string? s2, string? s3);
}

namespace System.Runtime.CompilerServices
{
  class EnsuresNotNullAttribute : System.Attribute { }
}
```

There's a warning for `s1.ToString()` in the second argument. I believe it's already handled in Roslyn by multiple passes over the method arguments (one that reports the diagnostics before the assertion, another with updated variable states that infers `s2` to be not null after the method call).

Given that regular assertion methods work exactly the same why Roslyn is unable to infer the type of `T` on the second pass?

**Comment by gafter** [comment]
The compiler does infer that the type of the parameter `s1` is non-null *after the method invocation*.  Since the possibly-null parameter is a parameter to the method invocation, type inference uses that type to infer the type argument to the method, resulting in a nullable return value for the method.
